vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/AndroidDevice.kt

```diff
@@ -164,26 +153,76 @@ class AndroidDevice(val ddmsDevice: IDevice,
 
         val timer = SystemTimer()
 
-        return CompositeTestRunListener(
+        testRunResultsListener = TestRunResultsListener(
+                testBatch = testBatch,
+                device = this,
+                timer = timer,
+                attachmentProviders = attachmentProviders)
+
+        val listeners = CompositeTestRunListener(
                 listOf(
                         recorderListener,
                         logCatListener,
-                        TestRunResultsListener(testBatch, this, deferred, timer, attachmentProviders),
+                        testRunResultsListener!!,
                         DebugTestRunListener(this),
                         ProgressTestRunListener(this, devicePoolId, progressReporter)
                 )
         )
+
+
+        val runner = async {
+            AndroidDeviceTestRunner(this@AndroidDevice)
+        }.await()
+
+        val deferredResult = async {
+            try {
+                runner.execute(configuration, testBatch, listeners)
+            } finally {
+                listeners.terminate()
+            }
+        }
+
+        val expectedTime: Long = testBatch
+                .calculateTimeout(configuration)
+                .run { this + this / 10L } // +10%
+
+        val expectedFinish = currentTimeMillis() + expectedTime
+
+        while (deferredResult.isActive) {
+            delay(1000)
```

Comment:
I planned to use channel instead of deferred to report status from android device to deviceActor, probably you need to take a look at this approach